src/buf_writer.c

```diff
@@ -37,6 +37,18 @@ buf_writer_init(buf_writer_t *buf_writer, write_cb_t *write_cb, void *cbopaque,
 	return buf_writer->buf == NULL;
 }
 
+static write_cb_t buf_writer_cb;
+
+write_cb_t *
+buf_writer_get_write_cb(buf_writer_t *buf_writer) {
+	return buf_writer->buf == NULL ? buf_writer->write_cb : buf_writer_cb;
+}
+
+void *
+buf_writer_get_cbopaque(buf_writer_t *buf_writer) {
+	return buf_writer->buf == NULL ? buf_writer->cbopaque : buf_writer;
```

Comment:
I’m lost here. Are we trying to provide some fall back options? Doesn’t seem right.

Reply:
Yes, that's part of the encapsulation: which call back function and which argument to use should be determined by the buffered writer module. If the caller requested a dynamic buffer and the buffered writer OOMed, the caller has a way to know it (from the return of `buf_writer_init()`), but the caller doesn't have to, and in such a case the caller can just safely use whatever function and argument the buffered writer gives back. See the usage examples as well as the unit tests.

Reply:
I guess the part I'm not quite follow, is why we return the `buf_writer` itself when it doesn't have any buffer (implies OOM?). Don't we always pass down the buf_writer always?

Reply:
We don't always pass down the buffered writer itself. In case of OOM, we fall back to the original (unbuffered) call back function and argument. The encapsulation means that this decision should be part of the buffered writer, and it's feasible but not required for the caller to handle it.